### ISSUE TITLE
[MICROGRANT]  Cara-Dune - mutliplayer _-like game with peer-to-peer IPFS networking

### ISSUE THREAD

**Issue by Sergei-Udris** [ipfs/devgrants#164]
### 1. What is your project? (max 100 words)

https://github.com/Sergei-Udris/Karga  
  - multiplayer game, Heroes III and Conquiztador combined - players will path find but instead of battles question challenges
  - IPFS networking - pubsub and circuit relay

i just finished release of his step 40 mins ago - and Karga is a payable game
- https://github.com/Sergei-Udris/Karga/releases/tag/24e48b7
- on needs java installed - i ran him on 16 and 11
- then download the jar - https://github.com/Sergei-Udris/Karga/releases/download/24e48b7/Karga-24e48b7.jar
  

### 2. How are you planning to improve this project? (max 200 words)

- i will add IPFS networking - use pubsub and circuit relay
- so players can join the same topic to play multiplayer
- i already did that in js-ipfs as part of previous adventures
- but now Karga will use jvm-libp2p for networking
- it was challenging but i managed to use jvm-libp2p from Clojure, i stopped at pubusub not working
- now with Karga will resolve it and mutliplayer will be there

### 3. Will the work be Open Source?

- Karga always will be free and open source
- no ads-donations-coins - he is a light in the room, another program to strengthen peer-to-peer future and IPFS
- he a cherry on top, a guest
- he will always be simple - not grow to much
- that is why i need the microgrant and not a macrogrant - to deliver an exquisit fuit to IPFS table, not a fast-food restaraunt unhappy-meal

### 4. If selected, do you agree to complete weekly updates and a grant report upon conclusion?

- i will continuously release Karga as he grows as whole program
- the report - the only report i seek is users actually playing the game regularly, that's the report
- and that will happen - because games, especially free with networking, let alone peer-to-peer with circuit relay  - are scarse
- Karga is correcting that
  
### 5. Does your proposal comply with our Community Code of Conduct?

- Karga is open source and i'm programming him as peer-to-peer multiplayer game
- so yes

### 6. Links and submissions

https://airtable.com/shrDZMizx03jOa4mQ

* If your project began at a hackathon, have you submitted it for the relevant Protocol Labs prizes? Include links here if available:
  no
* Have you filled out the [Interplanetary Builder Feedback Survey](https://airtable.com/shrDZMizx03jOa4mQ)? 
  yes https://airtable.com/shrDZMizx03jOa4mQ
### Additional questions:
* For each team member(s), please list name, email, Github account, and role in the project.
Karga is progammed by one person only - i am programmer
* If your project was created as part of an event or hackathon:
  * What was the name of the event? (e.g. ETHGlobal NFTHack, Cal Hacks hello:world, Chainlink, CivHacks, GameDevJ, ETHGlobal Scaling Ethereum)
  * Please link to your hackathon submission, proving you're a team that submitted for a Protocol Labs prize.

**Comment by Sergei-Udris** [comment]
- just tried again to make jvm-libp2p work - but encountered same situation - library is too immature/unready to be used
- specifically, pubsub - only works if two peers directly connected - but that's not what pubsub is for
- Karga, however, will be a multiplayer peer-to-peer Heroes 3 and Conquiztador game
- Karga will use torrent network  - not IPFS - will make use of BEP-50 for pubsub
- however
  - git and torrent are two fundamental concepts that IPFS implements
  - so i still want to apply for microgrant and see Karga join ipfs namespaces, even though using torrents
  - we are on the same side - distributed decentralized web3 and every program - every - that allows users to experience peer-to-peer - counts
  - Karga will not directly promote IPFS, but will promote values IPFS stands upon - open-source, peer-to-peer, free, torrent
- stay tuned

**Comment by Sergei-Udris** [comment]
- BEP-50 is a draft - pubsub over torrent does not exist
  - other networks - DAT, Scattlebutt, I2P - do not have pubsub
  - Tor is not peer-to-peer
  - jvm-libp2p is unripe and has 0 - zero - programs using it - otherwise where are links to them on github?
  - pubsub over go-ipfs or js-ipfs - does what Karga need, but program needs jvm
  - now - looking for solution - how to have that available on jvm as library for Karga

**Comment by Sergei-Udris** [comment]
Karga will use Mythrol for peer-to-peer pubusub networking https://github.com/Sergei-Udris/Karga/blob/main/deps.edn#L17
- Mythrol will be libp2p implementation for clojure
- Mythrol has two choices when it comes to how-he-will-do-it
- current and preferred - use jvm-libp2p and be a thin layer for use in clojure programs
  - requires figuring out why pubsub does not work in practice
  - as is also confirmed here https://discuss.ipfs.io/t/gossip-between-go-libp2p-jvm-libp2p/13599 
- or will gradually or abruptly implement libp2p in clojure - this is not desired, but i'm not afraid of the challenge
- bottom line
  - Karga needs peer-to-peer pubsub for  multiplayer available as library on jvm
  - Mythrol will provide that - and yes, Karga will use IPFS after all

**Comment by Sergei-Udris** [comment]
- no, it must be torrent
- will fugure out how

**Comment by Sergei-Udris** [comment]
networking solved
- all is already programmed an ready to use
- Karga will run alongside IPFS Desktop or ipfs daemon and use java-ipfs-http-client to connect and use pubsub
- to play the game user only needs to intall java, IPFS and download double-click Karga.jar
- programming multiplayer on one laptop works as well - run an ipfs node in a terminal for each player
- i already programmed and used pubsub in this manner using nodejs and js-ipfs-htt-client, all running within docker
- now - all is simplified as Karga is a dekstop program and will simply connect to an alongiside running IPFS node

i made several comments today - as networking was the biggest unknown of the game
- now it is fundamentally resolved
- so the next comment will be the release of Karga where we players can actually play multiplayer game

**Comment by Sergei-Udris** [comment]
Karga will have 7 tabs - processes - delivering more than just a game
- raisins - Conquiztador-like question challenge game
- sunflower-seeds - Heroes of Might and Magic III - like tile-map game
- apples - TBD
- salt - a unique vegan-themed game
- microwaved-potatoes - 
- corn - running simulations with OpenSCAD-like editor-on-the-left visualize-on-the-right, 2D
- beans - QCAD-LibreCAD 2D computer aided design - also with clojure-editor-on-the-left visualize-on-the-right

games
- will be multiplayer
- will have both mouse and programming language gameplay
- mouse is enough, but programming language will remove the performance cap for competitive play

as stated in previous comment - next is multiplayer for raisins process - Conquiztador question challenge course game

**Comment by Sergei-Udris** [comment]
twitch.tv users
- Karga's first, numerous and most hungry audience will be twitch.tv streamers and community around
- there was no release today but Karga now can exchange messages over IPFS pubsub and first multiplayer release is coming
- i will post here and then - most importantly - email to twitch streamers and comment under their Youtube videos that Karga seeks users
- out of 7 tabs - processes of the program  3 will be games, and having been  a twitch viewer for ages i know how streamers want new games, especially competitive multiplayer - to have new tournaments and liquipedia  and teamliquid
- players competing, streamers streaming and uploading videos on youtube - this is what Karga's game processes are for - food for hungry e-sports and gaming world of twitch

**Comment by Sergei-Udris** [comment]
https://github.com/Sergei-Udris/Cara-Dune/releases/tag/bd564f1 will take the relay race baton from Karga
- right this morning i removed Karga's gameplay as Conquiztador like game - but immediately reverted
- it is always better to not uproot a working program but start a new one if change is singinficant
- and - now understanding that game will be for twitch users - and the fact that i specifically decided not Heroes/Conquiztador but another game - this all means creating a new program and leaving Karga in his current state
- Cara-Dune continues right where Karga is - with working cross-platform GUI and message exchange over IPFS pubsub
- next - i will browse twitch.tv/directory and find a game or a couple Cara will port into 
  - being always open-source and free
  - with light 2D board-game like UI
  - without ads or anything else
  -  competitive, worthy of twtich streamers and liquipedia tournaments
  - without micro but focus on decision making
  - will take the best from Voobly - find join create games 
  - will not have tabs as program processes - but tabs where necessary for game-creating lobby ui

**Comment by Sergei-Udris** [comment]
Cara-Dune will have programming language gameplay
- editor on the left, 2D canvas to show the state of the game on the right
- twitch users are constantly - constantly - in search of new games that they can continuously stream and compete
- programming language gameplay is the new food - new kind of games - that has almost no gets-boring limitation - just like writing
- canvas on the right will vizualize some state and processes - processes - as multiple little waterfalls and program on the left will bring them in motion
- player's goal is to achieve a certain optimal state or unlock something by moving the mechanism
- multiplayer will show a tab for each player - or another canvas where we can choose a player and see their state
- Cara-Dune will have a host, players and observers - just open the any game's IPFS pubsub frequency and observe

release https://github.com/Sergei-Udris/Cara-Dune/releases/tag/650093f
to run install java, IPFS Desktop, download make-executable and double-click Cara-Dune.jar
- we can see editor on the left with a clojure core.async process and canvas on the right
- we evaluate this code and can see canvas changing as process steps forward

**Comment by Sergei-Udris** [comment]
cooking or assembling - programming libraries or not
- there's no point in cooking legumes or grains - that should be done by the houses, visitors should just choose and eat
- and there's quite a verity even in stores - canned beans, frozen corn and one can microwave potatoes
- same with libraries - instead of jumping to program proper libp2p on jvm program will run alongside IPFS Desktop or go-ipfs node
- however if there is no choice but - cook and be not deficient or not cook and be left without grain or legume - then programming a library is a must
- if and only if there is no other choice
- so at my own will i will always - always - use what's been cooked by houses to assemble food - program - for users
- if cooked beans or grains are just not there - they must be programmed to not be deficient and have a complete whole foods plant based vegan meal - even if it's not a dish but a lesser interview-bowl - so we don't go around blaming veganism as wrong while not eating properly for no reason

Cara-Dune
- will require editor and eval and expression selection to be programmed as Cursive and other hypocritical closed-sourced robberies are not an option
- and such will gradually be programmed - so user has true clojure and lisp experience and can play the game

**Comment by Sergei-Udris** [comment]
Cara-Dune's gameplay born, unfiltered unabridged unedited - gedit - version

we use simple shapes and colors
that then will become drawings of characters and map
but first - just shapes and clolros and processes and queues

channel orchestration
like making decisions whic to take from
which to put to
and what to put

start with the score?
simple as that
who scores mre
let's go
one simulation
tabs
score - just like Watney or Calican
an who scores more
and lobby
and final game table - who scored what

so one simulati0n with Cara

events 
iputs 
jsut events
writing processes - toherwise mess
there's a stream of events of all kinds - let be serveral
need to write several processes and carefully without breakage put ot proper channels

so there are processes and channels ready to receive events
player adds a few  Cara processes to take a stream of events and give to those processes meaniglfully
and - that stream comes over a number of channels
	
if destinations are stream of events - so it's not events that are moving and Cara standing
but Cara is moving between points on map - and those are values generated and put into channel

Cara is dropped to a new location with each event - value taken from channel
and she amkes adecision there - puts values on Mando's channel?
aha - she can choose from several channels of different location types which to process in whcih order

so she decides this way partially where to go
and recharges energy
goal - to bring Grogu to Jedi
so she puts values onto Mando's channel for him to move as well
Mando has Grogu
and they get to a point on the map where Mof Gideon cruiser is
so she helps Mando - and Mayfeld may be there - to get to the Cruiser and Grogu
that's it

**Comment by ErinOCon** [comment]
Hi @Sergei-Udris, Thank you for the submitted proposal - congrats on the progress made thus far! This microgrant has been approved. Look out for an email to submit payment details.
